### PR TITLE
[jit] Add default values in script

### PR DESCRIPTION
**Pull request by driazati**
PR adds support for `int` and `float` default values for script functions (values get cast to `Tensor` implicitly). Resolves expressions via `eval` (but only for types that can convert to a `Tensor`)

@apaszke @zdevito @ezyang

### CODE REVIEW

File: torch/csrc/jit/script/tree_views.h [hunk 1]
```diff
@@ -558,6 +477,105 @@ struct Const : public Expr {
   }
 };
 
+////////////////////////////////////////////////////////////////////////////////
+// Helper nodes (mostly for function arguments)
```
Comment (driazati) [inline torch/csrc/jit/script/tree_views.h]: Had to move this to use `Const` from inside `Param` to add `Param::has_default()` and `Param::default_value()`

File: torch/csrc/jit/ir.h [hunk 2]
```diff
@@ -171,6 +171,7 @@ using NodeKind = Symbol;
 struct Value {
   TH_DISALLOW_COPY_AND_ASSIGN(Value);
   Value(Node * node_, size_t offset_);
+  virtual ~Value() {}
```
Comment (zdevito) [inline torch/csrc/jit/ir.h]: There is no need for any changes to ir.h to add default values. We already support default values in `FunctionSchema` and each `script::Method` object has a function schema.

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -1931,9 +1937,24 @@ std::vector<Argument> parseArgsFromDecl(Decl decl, bool is_method) {
   size_t i = is_method ? 1 : 0;
   for (; i < decl.params().size(); ++i) {
     auto decl_arg = decl.params()[i];
-    auto arg = Argument(decl_arg.ident().name(), parseTypeFromExpr(decl_arg.type()),
-                        /*N =*/at::nullopt, /*default_value =*/at::nullopt,
-                        /*kwarg_only =*/false);
+    at::optional<IValue> default_value = at::nullopt;
+    if (decl_arg.has_default()) {
+      // If the Param has a default value, convert it to an IValue
+      Const default_const = decl_arg.default_value();
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: I am not sure what is going on here. I would rather not support default arguments from our string frontend yet, as it requires implementing a bunch more infrastructure than we currently have.

**Review by zdevito — CHANGES_REQUESTED** [review]
Adding support for default arguments is a great thing to work on, but I think this is ultimately not the right approach. Instead, I would suggest a much simpler set of changes:

1. Extract the default values from the original functions using: https://stackoverflow.com/questions/12627118/get-a-function-arguments-default-value
2. Compile the function into a graph as is, without changing the compiler or `ir.h`, deriving a `script::Method` object without any default values in it. Use the types inside of the derived `FunctionSchema`object and `inline IValue toIValue(py::handle obj, const TypePtr& type)` to turn the python default value into an IValue. 
3. Generate a new `FunctionSchema` object that includes the default IValues, and call `setSchema` on the Method to install it.

This approach is more modular (it doesn't require changes to AST parsing, the compiler, or the IR), and it relies on components we already have to robustly convert python values into IValue objects.

**Pull request closed without merging**